### DISCUSSION TITLE
Using FluidPatcher as MIDI Host

### DISCUSSION THREAD

**Discussion by derKarsten** [GeekFunkLabs/fluidpatcher#113] [Q&A]
Hi,
I'd like to connect a USB Midi synthesizer with a USB Midi controller. Can I use FluidPatcher in a fashion to ignore all messages from one Midi channel (i.e. 2) and just transfer them between these two other devices?

I'm really amazed from your software. I was looking for something similar for a while and am really happy to have found it.
 Best regards,
Karsten

**Comment by albedozero** [comment]
FluidPatcher uses FluidSynth's internal MIDI router to do its filtering, transposing, etc. of MIDI messages - but those routed messages can only go to FluidSynth. What you can do is use the `aconnect` command to manually connect your MIDI controller to your USB synth, and then set up your patches in FluidPatcher so there are no instruments on channel 2, and have your USB synth only respond to that channel.

**Reply by derKarsten** [reply]
Ok thanks, I'll look into that.

**Comment by pawelk** [comment]
I do this the way described by @albedozero , using `aconnect` [1] to connect all USB MIDI devices to each other. Basically all messages received on given channel from device A are sent on the same channel to devices B, C and so on and to Fluidsynth.

This way I can
1. set my MIDI devices to channels that are not utilized in Fluidpatcher bankfiles, so switching the channel on my keyboard plays either Fluidsynth or a hardware synth, or
2. set a hardware synth to share MIDI channel with fluidsynth patch, so I can layer e.g. fluidsynth piano with hardware synth drone

As far as I am aware Fluidsynth is a black hole for MIDI messages, i.e. there's no "midi-thru" kind of port that would output whatever messages were thrown at it, as is often the case with hardware synths. 

However, Fluidsynth docs state that:

> The MIDI driver API has a clean separation between the midi thread and the synthesizer. That opens the door to add a midi router module.

If it would be possible to tap into this, this would open up some fantastic possibilities, like e.g. creating a router rule that listens on channel 1, sends notes to fluidsynth instrument on channel 1, but also trasnposes an octave down and sends to e.g. channel 3 which can be picked up by whatever device(s) listening on that port.


[1] I actually use https://github.com/sinedied/pi-midi-host that calls `aconnect` autimatically when a device is plugged/unplugged. It does not clash with squishbox/fluidsynth midi auto connect feature.
